src/sinks/file/mod.rs

```diff
@@ -88,98 +95,137 @@ impl PartitionedFileSink {
 
         Some(bytes)
     }
-}
-
-impl Sink for PartitionedFileSink {
-    type SinkItem = Event;
-    type SinkError = ();
-
-    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        if let Some(key) = self.partition_event(&event) {
-            let encoding = self.encoding.clone();
 
-            let partition = self
-                .partitions
-                .entry(key.clone())
-                .or_insert_with(|| File::new(key.clone(), encoding.clone()));
-
-            self.last_accessed.insert(key.clone(), Instant::now());
+    fn deadline_at(&self) -> Instant {
+        Instant::now()
+            .checked_add(self.idle_timeout)
+            .expect("unable to compute next deadline")
+    }
 
-            partition
-                .start_send(event)
-                .map_err(|error| error!(message = "Error writing to partition.", %error))
-        } else {
-            Ok(AsyncSink::Ready)
+    async fn run(&mut self, input: impl Stream<Item = Event> + Send + Sync) -> crate::Result<()> {
+        pin_mut!(input);
+        loop {
+            tokio02::select! {
+                event = input.next() => {
+                    match event {
+                        None => {
+                            // If we got `None` - terminate the processing.
+                            debug!(message = "Receiver exausted, terminating the processing loop.");
+                            break;
+                        }
+                        Some(event) => self.process_event(event).await,
+                    }
+                }
+                result = self.files.next_expired(), if !self.files.is_empty() => {
+                    match result {
+                        None => unreachable!(),
+                        Some(Ok((mut expired_file, path))) => {
+                            // We got an expired file. All we really want is to
+                            // flush and close it.
+                            if let Err(error) = expired_file.flush().await {
+                                error!(
+                                    message = "Failed to flush file.",
+                                    path = ?path.get_ref(),
```

Comment:
it should borrow the path anyways, I don't think you need the `get_ref`?

Reply:
Good catch, thanks!